---
title: "Genome-wide diversity estimation"
---

```{r data-preparation}
# check if file exists
if (!file.exists(params$file_metadata)) {
    log4r::error(fn_logger, "Error: metadata is not found. Exited.")
    knitr::knit_exit()
}

# open metadata file
metadata <- data.table::fread(params$file_metadata)

# check column names
metadata_cnames <- c("id","dir_fasta","dir_gff","grouping")
if (!all(metadata_cnames %in% colnames(metadata))) {
    log4r::error(fn_logger, "Error: missing column names for metadata. Exited.")
    knitr::knit_exit()
}

# # create outdir for each grouping
# vct_grouping <- unique(metadata$grouping)
# for (group in vct_grouping) {
#     dir_grouping <- paste0(getwd(), "/", group, "/")
#     if (!dir.exists(dir_grouping)) {
#         dir.create(dir_grouping, recursive=T)
#     }
# }
```

```{r wgs-v1}
# create outdir
dir_wgs_fasta <- paste0(getwd(), "/wgs/fasta/")
if (!dir.exists(dir_wgs_fasta)) {
    dir.create(dir_wgs_fasta, recursive=T)
}

dir_wgs_gff <- paste0(getwd(), "/wgs/gff/")
if (!dir.exists(dir_wgs_gff)) {
    dir.create(dir_wgs_gff, recursive=T)
}

# copy all alignments to one folder
for (i in 1:nrow(metadata)) {
    # copy the fasta alignment
    fn_fasta_out <- paste0(dir_wgs_fasta, metadata$id[i], ".fa")
    system(paste("cp", metadata$dir_fasta[i], fn_fasta_out))

    # copy the gff
    fn_gff_out <- paste0(dir_wgs_gff, metadata$id[i])
    system(paste("cp", metadata$dir_gff[i], fn_gff_out))
}

# extract list of groupings
ls_grouping <- list()
for (i in unique(metadata$grouping)){
  vct_subset <- unlist(subset(metadata$id, metadata$grouping == i))
  ls_grouping <- c(ls_grouping, list(vct_subset))
}

# run PopGenome
wgs_data <- readData(dir_wgs_fasta, gffpath=dir_wgs_gff)
wgs_data <- set.populations(wgs_data, ls_grouping)
wgs_data <- neutrality.stats(wgs_data)
wgs_data  <- diversity.stats(wgs_data)
```

```{r wgs-v2, eval=FALSE}
# iterate over input alignments
for (group in vct_grouping) {
    # extract individuals from the respective group
    df_subset <- subset(metadata, grouping==group)
    
    # create outdir for WGS
    dir_wgs <- paste0(getwd(), "/", group, "/wgs/")
    if (!dir.exists(dir_wgs)) {
        dir.create(dir_wgs, recursive=T)
    }

    # MSA output
    fn_output_msa <- paste0(dir_wgs, group, ".fa")
    msa_fasta <- NULL

    # iterate over individuals
    for (i in 1:nrow(df_subset)) {
        # read the input FASTA
        fasta_aln <- seqinr::read.fasta(df_subset$dir_fasta[i])
        
        # check the number of individuals
        len_taxa <- length(fasta_aln)
        if (len_taxa != 1) {
            log4r::warn(fn_logger, paste("Warn:", len_taxa, "individuals are found for", df_subset$id[i]))
        }

        # add individual to the list
        msa_fasta <- c(msa_fasta, fasta_aln)
    }

    # save the MSA
    seqinr::write.fasta(sequences=msa_fasta, names=names(msa_fasta), file.out=fn_output_msa, nbchar=100)
}

```

```{r busco, eval=FALSE}
# iterate over input alignments

# requires update: iteration over grouping
for (i in 1:nrow(metadata)) {
    # create an output directory
    dir_busco_output <- paste0(getwd(), "/busco/")
    if (!dir.exists(dir_busco_output)) {
        dir.create(dir_busco_output, recursive=T)
    }

    # run busco
    f_run_busco(metadata$dir_fasta[i], dir_busco_output, params$lineage, "genome", params$thread, params$exe_busco)
}
```

```{r data-filtering, eval=FALSE}
for (i in 1:nrow(metadata)) {
    # open annotation file
    annotation <- data.table::fread(paste("grep -v '^#'", metadata$dir_gtf[i]))

    # to be continued
}
```